redhat-developer/gitops-operator
# Add controller to enable metrics/alerts for argocd instances

Adds a new controller to watch for the creation of ArgoCD instances and to add the configuration to enable the built-in metrics stack to scrape metrics from the new ArgoCD.  Also adds an alert rule to report ArgoCD applications that are out of sync.

### How to test by running the operator locally

Login to an openshift cluster from the command line

Download the operator-sdk binary from here: https://github.com/operator-framework/operator-sdk/releases/tag/v0.17.2


Check out the argocd-metrics branch from https://github.com/jopit/gitops-operator

cd gitops-operator
go mod vendor
kubectl apply -f deploy/crds
operator-sdk run --local --watch-namespace ""

Create a new argocd application in the openshift-gitops namespace with auto-sync turned off, do not sync the app.  It may take several minutes (I've seen it take over five minutes) but eventually an alert should show up on the overview page in the openshift admin console.

<img width="638" alt="Screen Shot 2021-02-18 at 11 41 48 AM" src="https://user-images.githubusercontent.com/29612372/108390882-16da8f00-71df-11eb-8fc6-cd2fe9fa83c1.png">

You can also go to the Monitoring>Metrics page in the admin console and enter a PromQL query, for example `sum(argocd_app_info{dest_namespace=~"default",health_status!=""}) by (sync_status)`


### Testing note by Shoubhik
Please validate that deleting up an ArgoCD instance also removes all associated rolebindings/roles and monitoring objects that were created.
